File: decomp/BattleShip-main/debug_tools/reloc_extract/reloc_extract.c
/*
 * reloc_extract — Standalone SSB64:RELOC file extractor.
 *
 * Reads baserom.us.z64 directly, locates a RELOC file by file_id using the
 * exact same logic as Torch's RelocFactory::parse, decompresses VPK0 if
 * needed, and writes the raw decompressed bytes to a file.
 *
 * Usage:
 *   reloc_extract <baserom.z64> <file_id> <output.bin>
 *
 * The output bytes are in N64 BIG-ENDIAN order — they are the same bytes that
 * Torch stores in the .o2r archive's mDecompressedData blob, BEFORE any
 * pass1 BSWAP32 or struct fixup is applied.  Compare this against the same
 * file extracted from the .o2r side to verify Torch's extraction is lossless.
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>

#include "../../torch/lib/libvpk0/vpk0.h"

/* Mirror RelocFactory.h constants */
#define RELOC_TABLE_ROM_ADDR    0x001AC870u
#define RELOC_FILE_COUNT        2132u
#define RELOC_TABLE_ENTRY_SIZE  12u
#define RELOC_TABLE_SIZE        ((RELOC_FILE_COUNT + 1u) * RELOC_TABLE_ENTRY_SIZE)
#define RELOC_DATA_START        (RELOC_TABLE_ROM_ADDR + RELOC_TABLE_SIZE)

static uint16_t read_be_u16(const uint8_t *p) {
    return (uint16_t)((p[0] << 8) | p[1]);
}

static uint32_t read_be_u32(const uint8_t *p) {
    return ((uint32_t)p[0] << 24) | ((uint32_t)p[1] << 16) |
           ((uint32_t)p[2] << 8)  |  (uint32_t)p[3];
}

int main(int argc, char **argv) {
    if (argc != 4) {
        fprintf(stderr, "usage: %s <baserom.z64> <file_id> <output.bin>\n", argv[0]);
        return 2;
    }

    const char *rom_path = argv[1];
    uint32_t    file_id  = (uint32_t)strtoul(argv[2], NULL, 0);
    const char *out_path = argv[3];

    if (file_id >= RELOC_FILE_COUNT) {
        fprintf(stderr, "error: file_id %u out of range (max %u)\n",
                file_id, RELOC_FILE_COUNT - 1);
        return 1;
    }

    /* --- Load ROM --- */
    FILE *fp = fopen(rom_path, "rb");
    if (!fp) {
        fprintf(stderr, "error: cannot open ROM '%s'\n", rom_path);
        return 1;
    }
    fseek(fp, 0, SEEK_END);
    long sz = ftell(fp);
    fseek(fp, 0, SEEK_SET);
    if (sz <= 0) {
        fprintf(stderr, "error: empty ROM\n");
        fclose(fp);
        return 1;
    }

    uint8_t *rom = (uint8_t *)malloc((size_t)sz);
    if (!rom) {
        fprintf(stderr, "error: oom\n");
        fclose(fp);
        return 1;
    }
    if (fread(rom, 1, (size_t)sz, fp) != (size_t)sz) {
        fprintf(stderr, "error: short read\n");
        free(rom);
        fclose(fp);
        return 1;
    }
    fclose(fp);

    /* z64 magic check */
    if (!(rom[0] == 0x80 && rom[1] == 0x37 && rom[2] == 0x12 && rom[3] == 0x40)) {
        fprintf(stderr, "error: not a z64 (big-endian) ROM (header: %02X %02X %02X %02X)\n",
                rom[0], rom[1], rom[2], rom[3]);
        free(rom);
        return 1;
    }

    /* --- Read this file's table entry and the next one --- */
    size_t table_off = RELOC_TABLE_ROM_ADDR + (size_t)file_id * RELOC_TABLE_ENTRY_SIZE;
    if (table_off + RELOC_TABLE_ENTRY_SIZE * 2 > (size_t)sz) {
        fprintf(stderr, "error: ROM too small for table entry %u\n", file_id);
        free(rom);
        return 1;
    }

    const uint8_t *te = rom + table_off;
    uint32_t first_word            = read_be_u32(te + 0);
    int      is_compressed         = (first_word >> 31) != 0;
    uint32_t data_offset           = first_word & 0x7FFFFFFFu;
    uint16_t reloc_intern          = read_be_u16(te + 4);
    uint16_t compressed_words      = read_be_u16(te + 6);
    uint16_t reloc_extern          = read_be_u16(te + 8);
    uint16_t decompressed_words    = read_be_u16(te + 10);

    uint32_t next_first_word       = read_be_u32(te + 12);
    uint32_t next_data_offset      = next_first_word & 0x7FFFFFFFu;

    uint32_t compressed_bytes      = (uint32_t)compressed_words   * 4u;
    uint32_t decompressed_bytes    = (uint32_t)decompressed_words * 4u;

    fprintf(stderr,
            "file_id=%u  compressed=%d  data_offset=0x%X  "
            "reloc_intern=0x%04X  reloc_extern=0x%04X  "
            "comp_bytes=%u  decomp_bytes=%u\n",
            file_id, is_compressed, data_offset,
            reloc_intern, reloc_extern,
            compressed_bytes, decompressed_bytes);

    /* --- Locate ROM data --- */
    size_t data_rom_addr = RELOC_DATA_START + data_offset;
    if (data_rom_addr + compressed_bytes > (size_t)sz) {
        fprintf(stderr, "error: ROM too small for file data\n");
        free(rom);
        return 1;
    }
    const uint8_t *file_data = rom + data_rom_addr;

    fprintf(stderr, "data_rom_addr=0x%zX (RELOC_DATA_START=0x%X + offset=0x%X)\n",
            data_rom_addr, (unsigned)RELOC_DATA_START, data_offset);

    /* --- Decompress / copy --- */
    uint8_t *out;
    uint32_t out_bytes;

    if (is_compressed) {
        out = (uint8_t *)malloc(decompressed_bytes);
        if (!out) {
            fprintf(stderr, "error: oom\n");
            free(rom);
            return 1;
        }
        uint32_t result = vpk0_decode(file_data, compressed_bytes,
                                      out, decompressed_bytes);
        if (result == 0) {
            fprintf(stderr, "error: VPK0 decode failed\n");
            free(out);
            free(rom);
            return 1;
        }
        out_bytes = decompressed_bytes;
    } else {
        out = (uint8_t *)malloc(decompressed_bytes);
        if (!out) {
            fprintf(stderr, "error: oom\n");
            free(rom);
            return 1;
        }
        memcpy(out, file_data, decompressed_bytes);
        out_bytes = decompressed_bytes;
    }

    /* --- Write output --- */
    FILE *of = fopen(out_path, "wb");
    if (!of) {
        fprintf(stderr, "error: cannot open output '%s'\n", out_path);
        free(out);
        free(rom);
        return 1;
    }
    if (fwrite(out, 1, out_bytes, of) != out_bytes) {
        fprintf(stderr, "error: short write\n");
        fclose(of);
        free(out);
        free(rom);
        return 1;
    }
    fclose(of);

    fprintf(stderr, "wrote %u bytes to %s\n", out_bytes, out_path);

    free(out);
    free(rom);
    return 0;
}

File: decomp/BattleShip-main/debug_tools/reloc_extract/reloc_extract.py
#!/usr/bin/env python3
"""
reloc_extract — Standalone SSB64:RELOC file extractor and verifier.

Reads baserom.us.z64 directly using the same logic as Torch's RelocFactory,
decompresses VPK0 if needed, and writes the raw decompressed bytes.

Use cases:
  - Extract a single file by file_id
  - Compare against the bytes loaded by the new port (after pass1 BSWAP32)
    to verify Torch's extraction is lossless

Usage:
  reloc_extract.py extract <baserom.z64> <file_id> <output.bin>
  reloc_extract.py diff <baserom.z64> <file_id> <port_dump.bin>
"""

import struct
import sys
from pathlib import Path

# ---------------------------------------------------------------------------
# ROM layout constants — must match torch/src/factories/ssb64/RelocFactory.h
# ---------------------------------------------------------------------------
RELOC_TABLE_ROM_ADDR    = 0x001AC870
RELOC_FILE_COUNT        = 2132
RELOC_TABLE_ENTRY_SIZE  = 12
RELOC_TABLE_SIZE        = (RELOC_FILE_COUNT + 1) * RELOC_TABLE_ENTRY_SIZE
RELOC_DATA_START        = RELOC_TABLE_ROM_ADDR + RELOC_TABLE_SIZE


# ---------------------------------------------------------------------------
# VPK0 decoder — ported from torch/lib/libvpk0/vpk0.c
# ---------------------------------------------------------------------------
class BitStream:
    __slots__ = ("data", "size", "pos", "bits", "avail")

    def __init__(self, data: bytes):
        self.data = data
        self.size = len(data)
        self.pos = 0
        self.bits = 0
        self.avail = 0

    def refill(self, need: int) -> None:
        while self.avail < need:
            if self.pos >= self.size:
                raise ValueError("VPK0: bitstream underflow")
            self.bits = (self.bits << 8) | self.data[self.pos]
            self.pos += 1
            self.avail += 8

    def read(self, n: int) -> int:
        self.refill(n)
        self.avail -= n
        return (self.bits >> self.avail) & ((1 << n) - 1)


class HuffNode:
    __slots__ = ("left", "right", "value")

    def __init__(self):
        self.left = None
        self.right = None
        self.value = 0


def build_tree(bs: BitStream) -> HuffNode:
    """Read a Huffman tree from the bitstream in prefix notation."""
    stack = []
    while True:
        bit = bs.read(1)
        if bit != 0 and len(stack) < 2:
            break
        node = HuffNode()
        if bit != 0:
            node.left = stack[-2]
            node.right = stack[-1]
            stack[-2:] = [node]
        else:
            node.value = bs.read(8)
            stack.append(node)
    return stack[0]


def tree_decode(bs: BitStream, root: HuffNode) -> int:
    n = root
    while n.left is not None:
        n = n.right if bs.read(1) else n.left
    return n.value


def vpk0_decode(src: bytes) -> bytes:
    """Decompress a VPK0 buffer. Returns decompressed bytes."""
    if len(src) < 9 or src[:4] != b"vpk0":
        raise ValueError("VPK0: bad magic")

    # Decompressed size (BE u32 at offset 4)
    dec_size = struct.unpack(">I", src[4:8])[0]

    # The bitstream starts at byte 4 (consumes the 4-byte size + 1-byte sample method)
    bs = BitStream(src[4:])
    bs.read(16)  # high half of size
    bs.read(16)  # low half of size
    sample_method = bs.read(8)

    offsets_tree = build_tree(bs)
    lengths_tree = build_tree(bs)

    # Use bytearray for the output buffer so we can index into it as the LZ77
    # back-reference source.
    out = bytearray()
    while len(out) < dec_size:
        flag = bs.read(1)
        if flag == 0:
            out.append(bs.read(8))
        else:
            if sample_method != 0:
                # Two-sample offset mode
                sub_offset = 0
                extra_bits = tree_decode(bs, offsets_tree)
                value = bs.read(extra_bits) if extra_bits else 0
                if value <= 2:
                    sub_offset = value + 1
                    extra_bits2 = tree_decode(bs, offsets_tree)
                    value = bs.read(extra_bits2) if extra_bits2 else 0
                # copy_src = out - value * 4 - sub_offset + 8
                copy_src_idx = len(out) - value * 4 - sub_offset + 8
            else:
                extra_bits = tree_decode(bs, offsets_tree)
                value = bs.read(extra_bits) if extra_bits else 0
                copy_src_idx = len(out) - value

            len_bits = tree_decode(bs, lengths_tree)
            length = bs.read(len_bits) if len_bits else 0

            for _ in range(length):
                out.append(out[copy_src_idx])
                copy_src_idx += 1

    return bytes(out[:dec_size])


# ---------------------------------------------------------------------------
# RELOC extractor
# ---------------------------------------------------------------------------
def read_table_entry(rom: bytes, file_id: int):
    if file_id >= RELOC_FILE_COUNT:
        raise ValueError(f"file_id {file_id} out of range (max {RELOC_FILE_COUNT - 1})")

    table_off = RELOC_TABLE_ROM_ADDR + file_id * RELOC_TABLE_ENTRY_SIZE
    if table_off + RELOC_TABLE_ENTRY_SIZE * 2 > len(rom):
        raise ValueError("ROM too small for table entry")

    te = rom[table_off:table_off + RELOC_TABLE_ENTRY_SIZE * 2]
    first_word         = struct.unpack(">I", te[0:4])[0]
    is_compressed      = (first_word >> 31) != 0
    data_offset        = first_word & 0x7FFFFFFF
    reloc_intern       = struct.unpack(">H", te[4:6])[0]
    compressed_words   = struct.unpack(">H", te[6:8])[0]
    reloc_extern       = struct.unpack(">H", te[8:10])[0]
    decompressed_words = struct.unpack(">H", te[10:12])[0]

    next_first_word    = struct.unpack(">I", te[12:16])[0]
    next_data_offset   = next_first_word & 0x7FFFFFFF

    return {
        "file_id": file_id,
        "is_compressed": is_compressed,
        "data_offset": data_offset,
        "reloc_intern": reloc_intern,
        "reloc_extern": reloc_extern,
        "compressed_bytes": compressed_words * 4,
        "decompressed_bytes": decompressed_words * 4,
        "next_data_offset": next_data_offset,
    }


def extract_file(rom: bytes, file_id: int) -> bytes:
    info = read_table_entry(rom, file_id)
    print(
        f"file_id={info['file_id']}  compressed={info['is_compressed']}  "
        f"data_offset=0x{info['data_offset']:X}  "
        f"reloc_intern=0x{info['reloc_intern']:04X}  "
        f"reloc_extern=0x{info['reloc_extern']:04X}  "
        f"comp_bytes={info['compressed_bytes']}  "
        f"decomp_bytes={info['decompressed_bytes']}",
        file=sys.stderr,
    )

    data_rom_addr = RELOC_DATA_START + info["data_offset"]
    print(
        f"data_rom_addr=0x{data_rom_addr:X} "
        f"(RELOC_DATA_START=0x{RELOC_DATA_START:X} + offset=0x{info['data_offset']:X})",
        file=sys.stderr,
    )

    if data_rom_addr + info["compressed_bytes"] > len(rom):
        raise ValueError("ROM too small for file data")

    file_data = rom[data_rom_addr:data_rom_addr + info["compressed_bytes"]]

    if info["is_compressed"]:
        decompressed = vpk0_decode(file_data)
        if len(decompressed) != info["decompressed_bytes"]:
            print(
                f"WARNING: vpk0 produced {len(decompressed)} bytes, "
                f"table says {info['decompressed_bytes']}",
                file=sys.stderr,
            )
    else:
        decompressed = file_data[:info["decompressed_bytes"]]

    return decompressed


# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------
def cmd_extract(argv):
    if len(argv) != 3:
        print("usage: reloc_extract.py extract <baserom.z64> <file_id> <output.bin>",
              file=sys.stderr)
        return 2

    rom_path = Path(argv[0])
    file_id  = int(argv[1], 0)
    out_path = Path(argv[2])

    rom = rom_path.read_bytes()
    if rom[:4] != bytes([0x80, 0x37, 0x12, 0x40]):
        print(f"warning: ROM header is not z64 ({rom[:4].hex()})", file=sys.stderr)

    data = extract_file(rom, file_id)
    out_path.write_bytes(data)
    print(f"wrote {len(data)} bytes to {out_path}", file=sys.stderr)
    return 0


def cmd_diff(argv):
    if len(argv) != 3:
        print("usage: reloc_extract.py diff <baserom.z64> <file_id> <port_dump.bin>",
              file=sys.stderr)
        return 2

    rom_path  = Path(argv[0])
    file_id   = int(argv[1], 0)
    port_path = Path(argv[2])

    rom = rom_path.read_bytes()
    rom_data = extract_file(rom, file_id)
    port_data = port_path.read_bytes()

    print()
    print(f"ROM-extracted (truth):  {len(rom_data)} bytes")
    print(f"port dump:              {len(port_data)} bytes")

    if len(rom_data) != len(port_data):
        print(f"SIZE MISMATCH: rom={len(rom_data)} port={len(port_data)}")

    cmp_len = min(len(rom_data), len(port_data))
    diffs = []
    for i in range(cmp_len):
        if rom_data[i] != port_data[i]:
            diffs.append(i)

    if not diffs and len(rom_data) == len(port_data):
        print("\nIDENTICAL — no differences found")
        return 0

    print(f"\n{len(diffs)} byte differences in first {cmp_len} bytes")

    # Print first 30 diffs with context
    print("\nFirst 30 differing bytes (offset: rom -> port):")
    for off in diffs[:30]:
        print(f"  0x{off:06X}: 0x{rom_data[off]:02X} -> 0x{port_data[off]:02X}")

    # Find runs (regions of contiguous diffs) for higher-level analysis
    runs = []
    if diffs:
        run_start = diffs[0]
        run_end = diffs[0]
        for d in diffs[1:]:
            if d == run_end + 1:
                run_end = d
            else:
                runs.append((run_start, run_end))
                run_start = d
                run_end = d
        runs.append((run_start, run_end))

    print(f"\n{len(runs)} contiguous diff runs:")
    for start, end in runs[:20]:
        size = end - start + 1
        print(f"  0x{start:06X} .. 0x{end:06X}  ({size} bytes)")
        # Hex dump of the run, both sides
        hex_rom  = " ".join(f"{b:02X}" for b in rom_data[start:end + 1])
        hex_port = " ".join(f"{b:02X}" for b in port_data[start:end + 1])
        print(f"    rom : {hex_rom}")
        print(f"    port: {hex_port}")

    return 1


def main(argv):
    if len(argv) < 1:
        print(__doc__)
        return 2
    cmd = argv[0]
    if cmd == "extract":
        return cmd_extract(argv[1:])
    elif cmd == "diff":
        return cmd_diff(argv[1:])
    else:
        print(f"unknown command: {cmd}", file=sys.stderr)
        print(__doc__)
        return 2


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))
